src/Streamly/Data/Fold.hs

```diff
@@ -6,61 +6,55 @@
+-- value. Therefore, a fold can also be thought of as a sink or a stream
+-- consumer.
 --
--- Using this representation multiple folds can be combined efficiently using
--- combinators; a stream can then be supplied to the combined fold and it would
--- distribute the input to constituent folds according to the composition.  For
--- example, an applicative composition distributes the same input to the
--- constituent folds and then combines the resulting fold outputs.  Similarly,
--- a partitioning combinator divides the input among constituent folds.
+-- A 'Fold' is in fact a data representation of a standard left fold ('foldl').
+-- Unlike the standard left folds, 'Fold's can terminate at any point e.g. the
+-- 'head' fold would terminate immediately after returning the head element.
 --
--- = Monoids
+-- 'Fold's can be combined efficiently using combinators provided in this
+-- module; a stream supplied to the combined fold is distributed to the
+-- constituent folds according to the behavior of the combinator.  For example,
+-- the 'tee' combinator distributes the input stream to two folds and then
+-- combines the resulting fold outputs.  Similarly, a 'partition' combinator
+-- divides the input among constituent folds.
 --
--- Monoids allow modular folding in general.  Most of the folds in this module
--- can be expressed using 'mconcat' and a suitable 'Monoid'. Instead of writing
--- folds we can write Monoids instead and turn them into folds.
+-- = Accumulators and Terminating Folds
 --
--- = Performance Notes
+-- Folds in this module can be classified in two categories viz. accumulators
+-- and terminating folds. Accumulators do not have a terminating condition,
+-- they run forever and consume the entire stream, for example the 'length'
+-- fold. Terminating folds have a terminating condition and can terminate
+-- without consuming the entire stream, for example, the 'head' fold.
 --
--- 'Fold' representation is more efficient than using streams when splitting
--- streams.  @Fold m a b@ can be considered roughly equivalent to a fold action
--- @m b -> t m a -> m b@ (where @t@ is a stream type and @m@ is a 'Monad').
--- Instead of using a 'Fold' type one could just use a fold action of the shape
--- @m b -> t m a -> m b@ for folding streams. However, multiple such actions
--- cannot be composed into a single fold function in an efficient manner.
--- Using the 'Fold' type we can efficiently split the stream across mutliple
--- folds because it allows the compiler to perform stream fusion optimizations.
+-- = Monoids
 --
--- On the other hand, transformation operations (e.g. 'Streamly.Prelude.map')
--- on stream types can be as efficient as transformations on 'Fold' (e.g.
--- 'Streamly.Internal.Data.Fold.lmap').
+-- Monoids allow generalized, modular folding.  The accumulators in this module
+-- can be expressed using 'mconcat' and a suitable 'Monoid'.  Instead of
```

Comment:
```suggestion
-- Monoids allow generalized, modular folding. The accumulators in this module
-- can be expressed using 'mconcat' and a suitable 'Monoid'. Instead of
```